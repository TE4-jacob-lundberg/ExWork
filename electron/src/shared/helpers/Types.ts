export interface ILink {
  url: string;
  label: string;
  type: string;
}

export interface IGameFormFields {
  title: string;
  abbreviation: string;
  image: string; 
  xAxis: string;
  yAxis: string;
  fileNames: string[];
}

export interface IGameFormErrors { [id: string]: string; }

export interface IGame {
  id: string;
  title: string;
  abbreviation: string | null;
  image: IGameImage;
  links: ILink[];
  fileNames: string[];
}

export interface IGameImage {
  url: string;
  bannerPos: IPosition;
}

export interface IPosition {
  x: string;
  y: string;
}
