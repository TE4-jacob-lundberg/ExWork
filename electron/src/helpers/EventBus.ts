/* eslint-disable @typescript-eslint/no-explicit-any */
const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

export { EventBus };
