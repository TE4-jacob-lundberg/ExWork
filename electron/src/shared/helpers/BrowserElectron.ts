// eslint-disable-next-line @typescript-eslint/no-explicit-any
export let electron: any;

if (process.env.ENVIRONMENT === 'test') electron = require('electron');
else electron = window.require('electron');
