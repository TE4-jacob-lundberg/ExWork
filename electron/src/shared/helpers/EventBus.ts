/* eslint-disable @typescript-eslint/no-explicit-any */
interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

export { EventBus };
