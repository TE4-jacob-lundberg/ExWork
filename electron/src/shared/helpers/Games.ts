import { standardGames, IGame } from '../constants/standardGames';

export class Games {

  public static all(): IGame[] {
    if (!localStorage.getItem('games')) localStorage.setItem('games', JSON.stringify(standardGames));

    return JSON.parse(localStorage.getItem('games')!) || [];
  }

  public static addNew(game: IGame): string {
    const savedGames: IGame[] = JSON.parse(localStorage.getItem('games')!);
    if (savedGames.some(sGame => sGame.id === game.id)) return 'DUPLICATE_ID';
    
    localStorage.setItem('games', JSON.stringify(savedGames.concat(game)));
    return '';
  }
}
