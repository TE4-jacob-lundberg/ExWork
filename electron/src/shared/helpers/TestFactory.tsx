/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

export function factory(component: JSX.Element): RenderResult {
  return render(component);
}

export function routerFactory(component: JSX.Element, options: any = { history: createBrowserHistory() }): RenderResult {
  return factory(
    <Router history={options.history}>
      {component}
    </Router>
  );
}
