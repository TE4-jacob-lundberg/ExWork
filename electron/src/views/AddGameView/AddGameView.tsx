import React from 'react';
import { useHistory } from 'react-router-dom';
import { useIndexedDB } from 'react-indexed-db';

import GameFormComponent from '../../shared/components/GameFormComponent';
import PageTitleComponent from '../../shared/components/PageTitleComponent';
import { routes } from '../../shared/constants/routes';

interface Props {}

const AddGameView: React.FC<Props> = function () {
  const history = useHistory();
  const db = useIndexedDB('games');

  return (
    <> 
      <PageTitleComponent title="Add a Game" />
      <GameFormComponent 
        onSubmit={(values): void => {
          db.add({
            ...values, 
            links: [],
            image: {
              url: values.image, 
              bannerPos: { 
                x: values.xAxis.toString(), 
                y: values.yAxis.toString(),
              },
            },
          }).then(resp => history.replace(routes.showGame.replace(':gameID', resp.toString())));
        }}
      />
    </>
  );
};

export default AddGameView;
