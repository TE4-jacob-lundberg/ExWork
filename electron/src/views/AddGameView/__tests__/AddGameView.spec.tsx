import React from 'react';

import { routerFactory } from '../../../shared/helpers/TestFactory';
import AddGameView from '../AddGameView';

jest.mock('react-indexed-db', () => ({
  useIndexedDB: (): object => {
    return {};
  },
}));

const testComponent = <AddGameView />;

describe('AddGameView.tsx', () => {
  it('renders correctly', () => {
    const { container } = routerFactory(testComponent);
    
    expect(container).toMatchSnapshot();
  });
});
