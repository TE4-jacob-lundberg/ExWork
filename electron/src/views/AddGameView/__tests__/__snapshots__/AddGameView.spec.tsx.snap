// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddGameView.tsx renders correctly 1`] = `
<div>
  <div
    class="css-vnxadl"
    data-testid="page-title"
  >
    <h1
      class="css-g7rarf"
    >
      Add a Game
    </h1>
  </div>
  <div
    class="css-o6jotw"
  >
    <form>
      <h1
        class="css-1vjmso2"
      >
        Game Information
      </h1>
      <section
        class="css-6a1pbd"
      >
        <label
          class="css-13m3dri"
          for="title"
        >
          Title*
        </label>
        <input
          class="css-1o0il4p"
          id="title"
          name="title"
          type="text"
          value=""
        />
        <label
          class="css-13m3dri"
          for="abbr"
        >
          Abbreviation
        </label>
        <input
          class="css-1o0il4p"
          id="abbr"
          name="abbreviation"
          type="text"
          value=""
        />
        <label
          class="css-13m3dri"
        >
          Game files (*.exe, *.app)
        </label>
        <input
          class="css-1o0il4p"
          name="fileNames"
          type="text"
          value=""
        />
      </section>
      <h1
        class="css-1vjmso2"
      >
        Image Settings
      </h1>
      <section
        class="css-6a1pbd"
      >
        <label
          class="css-13m3dri"
        >
          Image*
        </label>
        <input
          class="css-1o0il4p"
          type="file"
        />
        
      </section>
      <div
        class="css-11r1swr"
      >
        <button
          class="css-ucx151"
          data-testid="submit-form"
          type="submit"
        >
          Save
          <i
            class="material-icons"
          >
            save
          </i>
        </button>
      </div>
    </form>
    <div
      class="css-m2j5dl"
      x="50"
      y="0"
    />
  </div>
</div>
`;
